### PR TITLE
fix(api): Fix Custom Primary Key API Support

### PR DESCRIPTION
**Pull request by tylerjroach**
- [ ] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*

*How did you test these changes?*
(Please add a line here how the changes were tested)

*Documentation update required?*
- [ ] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [ ] Added Unit Tests
- [ ] Added Integration Tests
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tylerjroach** [comment]
Reviewers, please take a close look to ensure no unexpected behavior differences from changes made to AppSyncRequestFactory (DataStore) and AppSyncGraphQLRequestFactory (API).

File: aws-api/src/test/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestAndResponseCPKTest.kt [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+        assertFalse(response.hasErrors())
+        assertEquals(post.postId, response.data.postId)
+        assertEquals(post.title, response.data.title)
+        assertEquals(post.createdAt, response.data.createdAt)
+        assertEquals(post.rating, response.data.rating, 0.0)
+        assertNull(response.data.blog)
+    }
+}
+
+class CommentIdentifier(commentId: String) : ModelIdentifier<Comment>(commentId), Serializable
```
Comment (eeatonaws) [inline aws-api/src/test/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestAndResponseCPKTest.kt:401]: Could the `CommentIdentifier` from the `Comment` model be used instead?

File: aws-api/src/test/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestFactoryTest.java [hunk 2]
```diff
@@ -77,6 +78,33 @@ public void buildQueryFromClassAndId() throws JSONException {
         );
     }
 
+    /**
+     * Validate construction of a GraphQL query from a class and an object ID.
+     * @throws JSONException from JSONAssert.assertEquals
+     */
+    @Test
+    public void buildQueryFromClassAndModelIdentifier() throws JSONException {
+        // Arrange a hard-coded name/age as found int the expected data file.
```
Comment (eeatonaws) [inline aws-api/src/test/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestFactoryTest.java:87]:
```suggestion
        // Arrange a hard-coded name/age as found in the expected data file.
```

File: core/src/test/java/com/amplifyframework/core/model/ModelIdentifierTest.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
+package com.amplifyframework.core.model
```
Comment (eeatonaws) [inline core/src/test/java/com/amplifyframework/core/model/ModelIdentifierTest.kt:1]: Missing copyright.

File: testmodels/src/main/java/com/amplifyframework/testmodels/cpk/AmplifyModelProvider.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
+package com.amplifyframework.testmodels.cpk;
```
Comment (eeatonaws) [inline testmodels/src/main/java/com/amplifyframework/testmodels/cpk/AmplifyModelProvider.java:1]: Missing copyright.

**Review by eeatonaws — COMMENTED** [review]
Left a few minor comments and the new test model classes are missing the copyright header, but core logic LGTM.

File: aws-api/src/test/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestAndResponseCPKTest.kt [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+        val request: GraphQLRequest<Post> =
+            AppSyncGraphQLRequestFactory.buildQuery(Post::class.java, identifier)
+        val response = responseFactory.buildResponse(request, responseJson)
+
+        // THEN
+        JSONAssert.assertEquals(requestJson, request.content, true)
+        assertFalse(response.hasErrors())
+        assertEquals("p1", response.data.postId)
+        assertEquals("t1", response.data.title)
+        assertEquals(3.4, response.data.rating, 0.0)
```
[the remark this answers is not in the document]
Comment (tylerjroach) [reply aws-api/src/test/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestAndResponseCPKTest.kt:287]: Good eye. These were real requests/responses that I captured but it looks like I didn't capture request/response at the same time and send the same values each time. At the end of the day, these tests are verifying that the request is getting constructed properly, and that the response (which is a real response from server) is parsed properly. The two aren't necessarily together though.

File: aws-api/src/test/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestAndResponseCPKTest.kt (hunk 1 above)
Comment (tylerjroach) [reply aws-api/src/test/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestAndResponseCPKTest.kt:401]: Good catch! This was used before we determined that ModelIdentifiers will be added to all future models.

File: testmodels/src/main/java/com/amplifyframework/testmodels/cpk/AmplifyModelProvider.java (hunk 4 above)
Comment (tylerjroach) [reply testmodels/src/main/java/com/amplifyframework/testmodels/cpk/AmplifyModelProvider.java:1]: Hmm, I think that would be ok here. These are codegen and we don't add licenses to our codegen classes that typically go into customer classes. I don't necessarily want to touch codegen'ed files.

Comment (tylerjroach) [reply testmodels/src/main/java/com/amplifyframework/testmodels/cpk/AmplifyModelProvider.java:16]: On second thought, looks like we've added them on all the others so went ahead and added to all the new codegen files.

**Comment by tylerjroach** [comment]
@eeatonaws Addressed all comments. Thank you.

File: aws-api-appsync/src/main/java/com/amplifyframework/api/aws/GraphQLRequestHelper.java [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+                    Collections.singletonMap(appSyncOpType(op.type()), appSyncOpValue(op))
+            );
+        } else if (queryPredicate instanceof QueryPredicateGroup) {
+            QueryPredicateGroup qpg = (QueryPredicateGroup) queryPredicate;
+
+            if (QueryPredicateGroup.Type.NOT.equals(qpg.type())) {
+                try {
+                    return Collections.singletonMap("not", parsePredicate(qpg.predicates().get(0)));
+                } catch (IndexOutOfBoundsException exception) {
+                    throw new IllegalStateException(
```
[the remark this answers is not in the document]
Comment (tylerjroach) [reply aws-api-appsync/src/main/java/com/amplifyframework/api/aws/GraphQLRequestHelper.java]: This was a good catch, thank you. API category will now catch this and throw IllegalStateException as before and DataStore will catch and throw DataStoreException as before.

**Review by eeatonaws — APPROVED** [review]

**Review by mattcreaser — APPROVED** [review]

**Pull request merged**